File: accbot-ios/AccBot/Domain/Model/NotificationTemplateArgs.swift
import Foundation

/// Structured arguments for notification templates.
/// Stored as JSON in the `templateArgs` column - text is rendered at display time
/// using the current locale, so language switches re-render old notifications.
///
/// Numbers are stored as raw strings (Decimal description) and formatted at render time.
enum NotificationTemplateArgs: Codable, Equatable {
    case purchase(cryptoAmount: String, crypto: String, fiatAmount: String, fiat: String, price: String? = nil, scheduledAt: String? = nil, executedAt: String? = nil)
    case error(crypto: String, errorMessage: String)
    case lowBalance(exchangeName: String, fiat: String, balance: String, remainingDays: Int, connectionName: String? = nil)
    case withdrawalThreshold(amount: String, crypto: String, exchangeName: String, connectionName: String? = nil)
    case belowMinimum(crypto: String, purchaseAmount: String? = nil, fiat: String? = nil, minOrderSize: String? = nil)
    case networkRetry(crypto: String, exchangeName: String, connectionName: String? = nil)
    case missedPurchases(count: Int, crypto: String, exchangeName: String, connectionName: String? = nil)

    // MARK: - Render

    /// Render localized title + message for display using the current locale.
    func render() -> (title: String, message: String) {
        switch self {
        case .purchase(let cryptoAmount, let crypto, let fiatAmount, let fiat, let price, let scheduledAt, let executedAt):
            var base = String(localized: "Bought \(cryptoAmount) \(crypto) for \(fiatAmount) \(fiat)")
            if let price {
                base += " @ \(price) \(fiat)"
            }
            if let scheduled = scheduledAt, let executed = executedAt {
                return (
                    String(localized: "DCA Purchase Completed"),
                    String(localized: "Plan \(scheduled) → executed \(executed)")
                    + " · " + base
                )
            }
            return (
                String(localized: "DCA Purchase Completed"),
                base
            )
        case .error(let crypto, let errorMessage):
            return (
                String(localized: "DCA Failed"),
                String(localized: "Failed to buy \(crypto): \(errorMessage)")
            )
        case .lowBalance(let exchangeName, let fiat, let _, let remainingDays, let connectionName):
            let days = max(1, remainingDays)
            let label = Self.connectionLabel(exchangeName, connectionName)
            return (
                String(localized: "Low balance on \(label)"),
                String(localized: "~\(days) days of \(fiat) remaining for DCA")
            )
        case .withdrawalThreshold(let amount, let crypto, let exchangeName, let connectionName):
            let label = Self.connectionLabel(exchangeName, connectionName)
            return (
                String(localized: "Withdrawal Recommended"),
                String(localized: "You have accumulated \(amount) \(crypto) on \(label) - consider withdrawing to cold wallet")
            )
        case .belowMinimum(let crypto, let purchaseAmount, let fiat, let minOrderSize):
            if let purchaseAmount, let fiat, let minOrderSize {
                return (
                    String(localized: "DCA Failed"),
                    String(localized: "Failed to buy \(crypto): Amount \(purchaseAmount) \(fiat) below exchange minimum \(minOrderSize) \(fiat)")
                )
            }
            return (
                String(localized: "DCA Failed"),
                String(localized: "Amount below minimum for \(crypto)")
            )
        case .networkRetry(let crypto, let exchangeName, let connectionName):
            let label = Self.connectionLabel(exchangeName, connectionName)
            return (
                String(localized: "Network Error"),
                String(localized: "\(crypto) purchase on \(label) failed - no internet.")
            )
        case .missedPurchases(let count, let crypto, let exchangeName, let connectionName):
            let label = Self.connectionLabel(exchangeName, connectionName)
            return (
                String(localized: "Missed Purchases"),
                String(localized: "\(count) missed \(crypto) purchases on \(label) while offline.")
            )
        }
    }

    // MARK: - JSON Coding

    /// Build "exchangeName - connectionName" label, or just exchangeName when no connection name.
    private static func connectionLabel(_ exchangeName: String, _ connectionName: String?) -> String {
        if let name = connectionName, !name.isEmpty {
            return "\(exchangeName) - \(name)"
        }
        return exchangeName
    }

    private enum CodingKeys: String, CodingKey {
        case type
        case cryptoAmount, crypto, fiatAmount, fiat
        case errorMessage
        case exchangeName, remainingDays, balance
        case amount, price
        case purchaseAmount, minOrderSize
        case scheduledAt, executedAt
        case retryAt
        case count
        case connectionName
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .purchase(let cryptoAmount, let crypto, let fiatAmount, let fiat, let price, let scheduledAt, let executedAt):
            try container.encode("purchase", forKey: .type)
            try container.encode(cryptoAmount, forKey: .cryptoAmount)
            try container.encode(crypto, forKey: .crypto)
            try container.encode(fiatAmount, forKey: .fiatAmount)
            try container.encode(fiat, forKey: .fiat)
            try container.encodeIfPresent(price, forKey: .price)
            try container.encodeIfPresent(scheduledAt, forKey: .scheduledAt)
            try container.encodeIfPresent(executedAt, forKey: .executedAt)
        case .error(let crypto, let errorMessage):
            try container.encode("error", forKey: .type)
            try container.encode(crypto, forKey: .crypto)
            try container.encode(errorMessage, forKey: .errorMessage)
        case .lowBalance(let exchangeName, let fiat, let balance, let remainingDays, let connectionName):
            try container.encode("lowBalance", forKey: .type)
            try container.encode(exchangeName, forKey: .exchangeName)
            try container.encode(fiat, forKey: .fiat)
            try container.encode(balance, forKey: .balance)
            try container.encode(remainingDays, forKey: .remainingDays)
            try container.encodeIfPresent(connectionName, forKey: .connectionName)
        case .withdrawalThreshold(let amount, let crypto, let exchangeName, let connectionName):
            try container.encode("withdrawalThreshold", forKey: .type)
            try container.encode(amount, forKey: .amount)
            try container.encode(crypto, forKey: .crypto)
            try container.encode(exchangeName, forKey: .exchangeName)
            try container.encodeIfPresent(connectionName, forKey: .connectionName)
        case .belowMinimum(let crypto, let purchaseAmount, let fiat, let minOrderSize):
            try container.encode("belowMinimum", forKey: .type)
            try container.encode(crypto, forKey: .crypto)
            try container.encodeIfPresent(purchaseAmount, forKey: .purchaseAmount)
            try container.encodeIfPresent(fiat, forKey: .fiat)
            try container.encodeIfPresent(minOrderSize, forKey: .minOrderSize)
        case .networkRetry(let crypto, let exchangeName, let connectionName):
            try container.encode("networkRetry", forKey: .type)
            try container.encode(crypto, forKey: .crypto)
            try container.encode(exchangeName, forKey: .exchangeName)
            try container.encodeIfPresent(connectionName, forKey: .connectionName)
        case .missedPurchases(let count, let crypto, let exchangeName, let connectionName):
            try container.encode("missedPurchases", forKey: .type)
            try container.encode(count, forKey: .count)
            try container.encode(crypto, forKey: .crypto)
            try container.encode(exchangeName, forKey: .exchangeName)
            try container.encodeIfPresent(connectionName, forKey: .connectionName)
        }
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        switch type {
        case "purchase":
            self = .purchase(
                cryptoAmount: try container.decode(String.self, forKey: .cryptoAmount),
                crypto: try container.decode(String.self, forKey: .crypto),
                fiatAmount: try container.decode(String.self, forKey: .fiatAmount),
                fiat: try container.decode(String.self, forKey: .fiat),
                price: try container.decodeIfPresent(String.self, forKey: .price),
                scheduledAt: try container.decodeIfPresent(String.self, forKey: .scheduledAt),
                executedAt: try container.decodeIfPresent(String.self, forKey: .executedAt)
            )
        case "error":
            self = .error(
                crypto: try container.decode(String.self, forKey: .crypto),
                errorMessage: try container.decode(String.self, forKey: .errorMessage)
            )
        case "lowBalance":
            self = .lowBalance(
                exchangeName: try container.decode(String.self, forKey: .exchangeName),
                fiat: try container.decode(String.self, forKey: .fiat),
                balance: try container.decodeIfPresent(String.self, forKey: .balance) ?? "",
                remainingDays: try container.decode(Int.self, forKey: .remainingDays),
                connectionName: try container.decodeIfPresent(String.self, forKey: .connectionName)
            )
        case "withdrawalThreshold":
            self = .withdrawalThreshold(
                amount: try container.decode(String.self, forKey: .amount),
                crypto: try container.decode(String.self, forKey: .crypto),
                exchangeName: try container.decode(String.self, forKey: .exchangeName),
                connectionName: try container.decodeIfPresent(String.self, forKey: .connectionName)
            )
        case "belowMinimum":
            self = .belowMinimum(
                crypto: try container.decode(String.self, forKey: .crypto),
                purchaseAmount: try container.decodeIfPresent(String.self, forKey: .purchaseAmount),
                fiat: try container.decodeIfPresent(String.self, forKey: .fiat),
                minOrderSize: try container.decodeIfPresent(String.self, forKey: .minOrderSize)
            )
        case "networkRetry":
            self = .networkRetry(
                crypto: try container.decode(String.self, forKey: .crypto),
                exchangeName: try container.decode(String.self, forKey: .exchangeName),
                connectionName: try container.decodeIfPresent(String.self, forKey: .connectionName)
            )
        case "missedPurchases":
            self = .missedPurchases(
                count: try container.decode(Int.self, forKey: .count),
                crypto: try container.decode(String.self, forKey: .crypto),
                exchangeName: try container.decode(String.self, forKey: .exchangeName),
                connectionName: try container.decodeIfPresent(String.self, forKey: .connectionName)
            )
        default:
            throw DecodingError.dataCorruptedError(forKey: .type, in: container, debugDescription: "Unknown type: \(type)")
        }
    }

    // MARK: - Helpers

    func toJSON() -> String? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func fromJSON(_ json: String) -> NotificationTemplateArgs? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(NotificationTemplateArgs.self, from: data)
    }
}

File: accbot-android/app/src/main/java/com/accbot/dca/data/local/NotificationTemplateArgs.kt
package com.accbot.dca.data.local

import org.json.JSONObject

/**
 * Structured arguments for notification templates.
 * Stored as JSON in the `templateArgs` column – text is rendered at display time
 * using the current locale, so language switches re-render old notifications.
 *
 * Numbers are stored as raw strings (BigDecimal.toPlainString()) and formatted
 * via NumberFormatters at render time.
 */
sealed class NotificationTemplateArgs {

    abstract fun toJson(): String

    /** Successful purchase (filled immediately). */
    data class Purchase(
        val cryptoAmount: String,
        val crypto: String,
        val fiatAmount: String,
        val fiat: String,
        val price: String,
        val scheduledAtEpochMs: Long? = null,
        val executedAtEpochMs: Long? = null
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_PURCHASE)
            put("cryptoAmount", cryptoAmount)
            put("crypto", crypto)
            put("fiatAmount", fiatAmount)
            put("fiat", fiat)
            put("price", price)
            if (scheduledAtEpochMs != null) put("scheduledAtEpochMs", scheduledAtEpochMs)
            if (executedAtEpochMs != null) put("executedAtEpochMs", executedAtEpochMs)
        }.toString()
    }

    /** Purchase submitted but awaiting confirmation (PENDING status). */
    data class PurchasePending(
        val fiatAmount: String,
        val fiat: String,
        val crypto: String,
        val price: String,
        val scheduledAtEpochMs: Long? = null,
        val executedAtEpochMs: Long? = null
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_PURCHASE_PENDING)
            put("fiatAmount", fiatAmount)
            put("fiat", fiat)
            put("crypto", crypto)
            put("price", price)
            if (scheduledAtEpochMs != null) put("scheduledAtEpochMs", scheduledAtEpochMs)
            if (executedAtEpochMs != null) put("executedAtEpochMs", executedAtEpochMs)
        }.toString()
    }

    /** DCA error with structured crypto + errorMessage. */
    data class Error(
        val crypto: String,
        val errorMessage: String
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_ERROR)
            put("crypto", crypto)
            put("errorMessage", errorMessage)
        }.toString()
    }

    /** Low fiat balance warning. */
    data class LowBalance(
        val exchangeName: String,
        val fiat: String,
        val remainingDays: Double
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_LOW_BALANCE)
            put("exchangeName", exchangeName)
            put("fiat", fiat)
            put("remainingDays", remainingDays)
        }.toString()
    }

    /** Withdrawal threshold reached. */
    data class WithdrawalThreshold(
        val amount: String,
        val crypto: String,
        val exchangeName: String
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_WITHDRAWAL_THRESHOLD)
            put("amount", amount)
            put("crypto", crypto)
            put("exchangeName", exchangeName)
        }.toString()
    }

    /** Target accumulation reached – plan auto-disabled. */
    data class TargetReached(
        val targetAmount: String,
        val crypto: String
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_TARGET_REACHED)
            put("targetAmount", targetAmount)
            put("crypto", crypto)
        }.toString()
    }

    /** Purchase amount below exchange minimum order size. */
    data class BelowMinimum(
        val crypto: String,
        val purchaseAmount: String,
        val fiat: String,
        val minOrderSize: String
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_BELOW_MINIMUM)
            put("crypto", crypto)
            put("purchaseAmount", purchaseAmount)
            put("fiat", fiat)
            put("minOrderSize", minOrderSize)
        }.toString()
    }

    /** Network error – purchase will be retried. */
    data class NetworkRetry(
        val crypto: String,
        val exchangeName: String,
        val errorMessage: String,
        val nextRetryAtEpochMs: Long,
        val attemptCount: Int,
        val planId: Long
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_NETWORK_RETRY)
            put("crypto", crypto)
            put("exchangeName", exchangeName)
            put("errorMessage", errorMessage)
            put("nextRetryAtEpochMs", nextRetryAtEpochMs)
            put("attemptCount", attemptCount)
            put("planId", planId)
        }.toString()
    }

    /** Missed purchases due to prolonged offline period. */
    data class MissedPurchases(
        val crypto: String,
        val exchangeName: String,
        val missedCount: Int,
        val planId: Long
    ) : NotificationTemplateArgs() {
        override fun toJson(): String = JSONObject().apply {
            put(KEY_TYPE, TYPE_MISSED_PURCHASES)
            put("crypto", crypto)
            put("exchangeName", exchangeName)
            put("missedCount", missedCount)
            put("planId", planId)
        }.toString()
    }

    companion object {
        private const val KEY_TYPE = "type"
        private const val TYPE_PURCHASE = "purchase"
        private const val TYPE_PURCHASE_PENDING = "purchase_pending"
        private const val TYPE_ERROR = "error"
        private const val TYPE_LOW_BALANCE = "low_balance"
        private const val TYPE_WITHDRAWAL_THRESHOLD = "withdrawal_threshold"
        private const val TYPE_TARGET_REACHED = "target_reached"
        private const val TYPE_BELOW_MINIMUM = "below_minimum"
        private const val TYPE_NETWORK_RETRY = "network_retry"
        private const val TYPE_MISSED_PURCHASES = "missed_purchases"

        fun fromJson(json: String): NotificationTemplateArgs? = try {
            val obj = JSONObject(json)
            when (obj.getString(KEY_TYPE)) {
                TYPE_PURCHASE -> Purchase(
                    cryptoAmount = obj.getString("cryptoAmount"),
                    crypto = obj.getString("crypto"),
                    fiatAmount = obj.getString("fiatAmount"),
                    fiat = obj.getString("fiat"),
                    price = obj.getString("price"),
                    scheduledAtEpochMs = obj.optLong("scheduledAtEpochMs", 0L).takeIf { it > 0 },
                    executedAtEpochMs = obj.optLong("executedAtEpochMs", 0L).takeIf { it > 0 }
                )
                TYPE_PURCHASE_PENDING -> PurchasePending(
                    fiatAmount = obj.getString("fiatAmount"),
                    fiat = obj.getString("fiat"),
                    crypto = obj.getString("crypto"),
                    price = obj.getString("price"),
                    scheduledAtEpochMs = obj.optLong("scheduledAtEpochMs", 0L).takeIf { it > 0 },
                    executedAtEpochMs = obj.optLong("executedAtEpochMs", 0L).takeIf { it > 0 }
                )
                TYPE_ERROR -> Error(
                    crypto = obj.getString("crypto"),
                    errorMessage = obj.getString("errorMessage")
                )
                TYPE_LOW_BALANCE -> LowBalance(
                    exchangeName = obj.getString("exchangeName"),
                    fiat = obj.getString("fiat"),
                    remainingDays = obj.getDouble("remainingDays")
                )
                TYPE_WITHDRAWAL_THRESHOLD -> WithdrawalThreshold(
                    amount = obj.getString("amount"),
                    crypto = obj.getString("crypto"),
                    exchangeName = obj.getString("exchangeName")
                )
                TYPE_TARGET_REACHED -> TargetReached(
                    targetAmount = obj.getString("targetAmount"),
                    crypto = obj.getString("crypto")
                )
                TYPE_BELOW_MINIMUM -> BelowMinimum(
                    crypto = obj.getString("crypto"),
                    purchaseAmount = obj.getString("purchaseAmount"),
                    fiat = obj.getString("fiat"),
                    minOrderSize = obj.getString("minOrderSize")
                )
                TYPE_MISSED_PURCHASES -> MissedPurchases(
                    crypto = obj.getString("crypto"),
                    exchangeName = obj.getString("exchangeName"),
                    missedCount = obj.getInt("missedCount"),
                    planId = obj.optLong("planId", 0)
                )
                TYPE_NETWORK_RETRY -> NetworkRetry(
                    crypto = obj.getString("crypto"),
                    exchangeName = obj.getString("exchangeName"),
                    errorMessage = obj.getString("errorMessage"),
                    nextRetryAtEpochMs = obj.getLong("nextRetryAtEpochMs"),
                    attemptCount = obj.optInt("attemptCount", 1),
                    planId = obj.optLong("planId", 0)
                )
                else -> null
            }
        } catch (_: Exception) {
            null
        }
    }
}
